sw/host/opentitanlib/src/transport/hyperdebug/dfu.rs

```diff
@@ -0,0 +1,433 @@
+impl Transport for HyperdebugDfu {
+    fn capabilities(&self) -> Result<Capabilities> {
+        Ok(Capabilities::new(Capability::NONE))
+    }
+
+    fn dispatch(&self, action: &dyn Any) -> Result<Option<Box<dyn Annotate>>> {
+        if let Some(update_firmware_action) = action.downcast_ref::<UpdateFirmware>() {
+            update_firmware(
+                &mut self.usb_backend.borrow_mut(),
+                &update_firmware_action.firmware,
+                &update_firmware_action.progress,
+            )
+        } else {
+            bail!(TransportError::UnsupportedOperation)
+        }
+    }
+}
+
+const USB_CLASS_APP: u8 = 0xFE;
+const USB_SUBCLASS_DFU: u8 = 0x01;
+
+const DFUSE_ERASE_PAGE: u8 = 0x41;
+const DFUSE_PROGRAM_PAGE: u8 = 0x21;
+
+const DFU_STATUS_OK: u8 = 0x00;
+
+const DFU_STATE_APP_IDLE: u8 = 0x00;
+const DFU_STATE_DFU_IDLE: u8 = 0x02;
+const DFU_STATE_DOWNLOAD_BUSY: u8 = 0x04;
+const DFU_STATE_DOWNLOAD_IDLE: u8 = 0x05;
+
+#[cfg(not(feature = "include_hyperdebug_firmware"))]
+const OFFICIAL_FIRMWARE: Option<&'static [u8]> = None;
+#[cfg(feature = "include_hyperdebug_firmware")]
+const OFFICIAL_FIRMWARE: Option<&'static [u8]> = Some(include_bytes!(env!("hyperdebug_firmware")));
+
+pub fn official_firmware_version() -> Result<Option<&'static str>> {
+    if let Some(fw) = OFFICIAL_FIRMWARE {
+        Ok(Some(get_hyperdebug_firmware_version(fw)?))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Helper method to verify that the given binary image looks like a HyperDebug firmware image.
+fn validate_firmware_image(firmware: &[u8]) -> Result<()> {
+    get_hyperdebug_firmware_version(firmware)?;
+    Ok(())
+}
+
+const EC_COOKIE: [u8; 4] = [0x99, 0x88, 0x77, 0xce];
+const EC_FIRMWARE_NAME_LEN: usize = 32;
+
+fn get_hyperdebug_firmware_version(firmware: &[u8]) -> Result<&str> {
+    let Some(pos) = firmware[0..1024]
+        .chunks(4)
+        .position(|c| c[0..4] == EC_COOKIE) else {
+            bail!(TransportError::FirmwareProgramFailed(
+                "File is not a HyperDebug firmware image".to_string()
+            ));
+        };
+    let firmware_name_field = &firmware[(pos + 1) * 4..(pos + 1) * 4 + EC_FIRMWARE_NAME_LEN];
+    let end = firmware_name_field
+        .iter()
+        .rev()
+        .position(|b| *b != 0x00)
+        .map(|j| EC_FIRMWARE_NAME_LEN - j)
+        .unwrap_or(0);
+    Ok(std::str::from_utf8(&firmware_name_field[0..end])?)
+}
+
+/// Helper method to perform flash programming using ST's DfuSe variant of the DFU protocol.
+/// This method is used both by the `Hyperdebug` and the `HyperdebugDfu` structs.
+pub fn update_firmware(
+    usb_device: &mut UsbBackend,
+    firmware: &Option<Vec<u8>>,
+    progress: &Option<Box<dyn Fn(Progress)>>,
+) -> Result<Option<Box<dyn Annotate>>> {
+    let firmware: &[u8] = if let Some(vec) = firmware.as_ref() {
+        validate_firmware_image(vec)?;
+        vec
+    } else {
+        OFFICIAL_FIRMWARE.unwrap()
```

Comment:
Maybe emit an error here if OFFICIAL_FIRMWARE is None:
```
OFFICIAL_FIRMWARE.ok_or(anyhow!("No built-in firmware"))?
```

Reply:
Done